activerecord/test/cases/adapters/mysql2/virtual_column_test.rb

```diff
@@ -0,0 +1,59 @@
+require "cases/helper"
+require "support/schema_dumping_helper"
+
+if ActiveRecord::Base.connection.supports_virtual_columns?
+  class Mysql2VirtualColumnTest < ActiveRecord::Mysql2TestCase
+    include SchemaDumpingHelper
+
+    self.use_transactional_tests = false
+
+    class VirtualColumn < ActiveRecord::Base
+    end
+
+    def setup
+      @connection = ActiveRecord::Base.connection
+      @connection.create_table :virtual_columns, force: true do |t|
+        t.string  :name
+        t.virtual :upper_name,  type: :string,  as: "UPPER(name)"
+        t.virtual :name_length, type: :integer, as: "LENGTH(name)", stored: true
+      end
+      VirtualColumn.create(name: "Rails")
+    end
+
+    def teardown
+      @connection.drop_table :virtual_columns, if_exists: true
+      VirtualColumn.reset_column_information
+    end
+
+    def test_virtual_column
+      column = VirtualColumn.columns_hash["upper_name"]
+      assert column.virtual?
+      assert_match %r{\bVIRTUAL\b}, column.extra
+      assert_equal "RAILS", VirtualColumn.take.upper_name
+    end
+
+    def test_stored_column
+      column = VirtualColumn.columns_hash["name_length"]
+      assert column.virtual?
+      assert_match %r{\b(?:STORED|PERSISTENT)\b}, column.extra
+      assert_equal 5, VirtualColumn.take.name_length
+    end
+
+    def test_change_table
+      @connection.change_table :virtual_columns do |t|
+        t.virtual :lower_name, type: :string, as: "LOWER(name)"
+      end
+      VirtualColumn.reset_column_information
+      column = VirtualColumn.columns_hash["lower_name"]
+      assert column.virtual?
+      assert_match %r{\bVIRTUAL\b}, column.extra
+      assert_equal "rails", VirtualColumn.take.lower_name
+    end
+
+    def test_schema_dumping
```

Comment:
It still fails with MySQL 8.0.0.

```ruby
$ ARCONN=mysql2 bundle exec ruby -w -Itest test/cases/adapters/mysql2/virtual_column_test.rb
/home/yahonda/git/rails/activesupport/lib/active_support/core_ext/enumerable.rb:20: warning: method redefined; discarding old sum
Using mysql2
Run options: --seed 3987

# Running:

...F

Finished in 1.154189s, 3.4656 runs/s, 12.1297 assertions/s.

  1) Failure:
Mysql2VirtualColumnTest#test_schema_dumping [test/cases/adapters/mysql2/virtual_column_test.rb:55]:
Expected /t\.virtual\s+"upper_name",\s+type: :string,\s+as: "UPPER\(name\)"$/i to match "# This file is auto-generated from the current state of the database. Instead\n# of editing this file, please use the migrations feature of Active Record to\n# incrementally modify your database, and then regenerate this schema definition.\n#\n# Note that this schema.rb definition is the authoritative source for your\n# database schema. If you need to create the application database on another\n# system, you should be using db:schema:load, not running all the migrations\n# from scratch. The latter is a flawed and unsustainable approach (the more migrations\n# you'll amass, the slower it'll run and the greater likelihood for issues).\n#\n# It's strongly recommended that you check this file into your version control system.\n\nActiveRecord::Schema.define(version: 0) do\n\n  create_table \"virtual_columns\", force: :cascade, options: \"ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci\" do |t|\n    t.string \"name\"\n    t.virtual \"upper_name\", type: :string, as: \"upper(`name`)\"\n    t.virtual \"name_length\", type: :integer, as: \"length(`name`)\", stored: true\n  end\n\nend\n".

4 runs, 14 assertions, 1 failures, 0 errors, 0 skips
```

Here "name" inside the upper and length function needs backquote.